src/canister_tests/src/tests.rs

```diff
@@ -66,6 +66,165 @@ fn registration_with_mismatched_sender_fails() {
     );
 }
 
+/// Tests for the HTTP interactions according to the HTTP gateway spec: https://internetcomputer.org/docs/current/references/ic-interface-spec/#http-gateway
+#[cfg(test)]
+mod http_tests {
+    use crate::certificate_validation::validate_certification;
+    use crate::framework::CallError;
+    use crate::{api, framework};
+    use ic_state_machine_tests::StateMachine;
+    use internet_identity_interface::{HeaderField, HttpRequest};
+    use itertools::Itertools;
+    use regex::Regex;
+    use serde_bytes::ByteBuf;
+
+    #[test]
+    fn ii_canister_serves_http_assets() -> Result<(), CallError> {
+        let assets: Vec<(&str, Option<&str>)> = vec![
+            ("/", None),
+            ("/index.html", None),
+            ("/index.js", Some("gzip")),
+            ("/loader.webp", None),
+            ("/favicon.ico", None),
+            ("/ic-badge.svg", None),
+        ];
+        let env = StateMachine::new();
+        let canister_id = framework::install_ii_canister(&env, framework::II_WASM_PREVIOUS.clone());
+
+        for (asset, encoding) in assets {
+            let http_response = api::http_request(
+                &env,
+                canister_id,
+                HttpRequest {
+                    method: "GET".to_string(),
+                    url: asset.to_string(),
+                    headers: vec![],
+                    body: ByteBuf::new(),
+                },
+            )?;
+
+            assert_eq!(http_response.status_code, 200);
+
+            // check the appropriate Content-Encoding header is set
+            if let Some(enc) = encoding {
+                let (_, content_encoding) = http_response
+                    .headers
+                    .iter()
+                    .filter(|(name, _)| name == "Content-Encoding")
+                    .exactly_one()
+                    .expect("error retrieving \"Content-Encoding\" header");
+                assert_eq!(
+                    content_encoding, enc,
+                    "unexpected \"Content-Encoding\" header value"
+                );
+            }
+
+            let (_, ic_certificate) = http_response
+                .headers
+                .iter()
+                .filter(|(name, _)| name == "IC-Certificate")
+                .exactly_one()
+                .unwrap();
+
+            validate_certification(
+                ic_certificate,
+                canister_id,
+                asset,
+                &http_response.body,
+                None, // should really be `encoding`, but we need to solve L2-722 first
+                env.root_key(),
+            )
+            .expect(&format!("validation for asset \"{}\" failed", asset));
+
+            verify_security_headers(&http_response.headers)
+        }
+        Ok(())
+    }
+
+    fn verify_security_headers(headers: &Vec<HeaderField>) {
```

Comment:
It'd be nice to keep this file as short as possible and only contain tests so that they're easy to read and get an idea of what's getting testing. Do you think we could put these helpers somewhere else?